target/riscv/cpu_helper.c

```diff
@@ -681,8 +696,24 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
                 *prot |= PAGE_WRITE;
             }
 #if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
-            if ((pte & PTE_LC) == 0)
-                *prot |= PAGE_LC_CLEAR;
+            if ((pte & PTE_CR) == 0) {
+                if ((pte & PTE_CRM) == 0) {
+                    *prot |= PAGE_LC_CLEAR;
+                } else {
+                    *prot |= PAGE_LC_TRAP;
+                }
+            } else {
+                if (pte & PTE_CRM) {
+                    // Cap-loads checked against GCLG in CCSR using PTE_U
+
+                    target_ulong gclgmask =
+                            (pte & PTE_U) ? CCSR_GCLGU : CCSR_GCLGS;
+
+                    if (!(env->sccsr & gclgmask) != !(pte & PTE_CRG)) {
```

Comment:
I know what you're trying to do but it's hard to understand mentally. I'd suggest shifting and masking the low bit instead.